pkg/provision/cleanup.go

```diff
@@ -0,0 +1,230 @@
+		log.Info().Msgf("Checking namespace %s", ns)
+		for _, cnt := range lines(must(execute("ctr", "-n", ns, "container", "ls", "-q"))) {
+			bytes := must(execute("ctr", "-n", ns, "container", "info", cnt))
+			var container container
+			if err := json.Unmarshal(bytes, &container); err != nil {
+				panic(err)
+			}
+
+			zdbCl := zdbConnection(container.ID)
+			defer zdbCl.Close()
+			if err := zdbCl.Connect(); err != nil {
+				return errors.Wrapf(err, "failed to connect to 0-db: %s", container.ID)
+			}
+
+			ns, err := zdbCl.Namespaces()
+			if err != nil {
+				return errors.Wrap(err, "failed to retrieve zdb namespaces")
+			}
+
+			toSave[filepath.Base(container.Spec.Root.Path)] = struct{}{}
+			for _, mnt := range container.Spec.Mounts {
+				if strings.HasPrefix(mnt.Source, "/mnt/") {
+					log.Info().Msgf("ZDB namespaces length: %d in path %s", len(ns), mnt.Source)
+					if len(ns) == 1 && ns[0] == "default" {
+						err := deleteZdbContainer(pkg.ContainerID(container.ID))
+						if err != nil {
+							return errors.Wrapf(err, "failed to delete zdb container %s", container.ID)
+						}
+						toDelete[filepath.Base(mnt.Source)] = struct{}{}
+					} else {
+						toSave[filepath.Base(mnt.Source)] = struct{}{}
+					}
+				}
+			}
+		}
+	}
+
+	ctx := context.TODO()
+	utils := filesystem.NewUtils()
+	pools, err := ioutil.ReadDir("/mnt")
+	if err != nil {
+		return errors.Wrap(err, "failed to list available pools")
+	}
+
+	for _, pool := range pools {
+		if !pool.IsDir() {
+			continue
+		}
+		path := filepath.Join("/mnt", pool.Name())
+		subvols, err := utils.SubvolumeList(ctx, path)
+		if err != nil {
+			// A Pool might not be mounted here, skip it
+			log.Error().Err(err).Msgf("failed to list available subvolumes in: %s", path)
+			continue
+		}
+
+		qgroups, err := utils.QGroupList(ctx, path)
+		if err != nil {
+			return errors.Wrapf(err, "failed to list available cgroups in: %s", path)
+		}
+
+		for _, subvol := range subvols {
+			qgroup, ok := qgroups[fmt.Sprintf("0/%d", subvol.ID)]
+			if !ok {
+				log.Info().Msgf("skipping volume '%s' has no assigned quota", subvol.Path)
+				continue
+			}
+
+			// now, is this subvol in one of the toDelete ?
+			if _, ok := toDelete[filepath.Base(subvol.Path)]; ok {
+				// delete the subvolume
+				log.Info().Msgf("deleting subvolume %s", subvol.Path)
+				if err := utils.SubvolumeRemove(ctx, filepath.Join(path, subvol.Path)); err != nil {
+					log.Err(err).Msgf("failed to delete subvol '%s'", subvol.Path)
+				}
+				if err := utils.QGroupDestroy(ctx, qgroup.ID, path); err != nil {
+					log.Err(err).Msgf("failed to delete qgroup: '%s'", qgroup.ID)
+				}
+				continue
+			}
+
+			// we only handle volumes that are 256MiB or 10MiB
+			if qgroup.MaxRfer != 268435456 && qgroup.MaxRfer != 10485760 {
+				// if the subvolume is a zdb and has 0 maxRfer, don't skip here. It might need to be deleted
+				if !strings.HasPrefix(subvol.Path, "zdb") && qgroup.MaxRfer != 0 {
```

Comment:
zdb volumes never have size limit as far as I remember. I don't also get what you skip zdb subvolumes that has size limit ?